File: deepseek_gateway.js
#!/usr/bin/env node
const http = require('http');
const https = require('https');
const { URL } = require('url');

const PORT = 5000;
const DEEPSEEK_BASE_URL = 'https://api.deepseek.com';
const THINKING_MODELS = ['deepseek-v4-pro', 'deepseek-v4-flash'];

const server = http.createServer((req, res) => {
    console.log(`${new Date().toISOString()} ${req.method} ${req.url}`);
    
    if (req.url === '/health' || req.url === '/v1/models') {
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ status: 'ok' }));
        return;
    }

    if ((req.url === '/' || req.url.startsWith('/v1/chat/completions')) && req.method === 'POST') {
        let body = '';
        
        req.on('data', chunk => { body += chunk.toString(); });
        req.on('end', () => {
            try {
                let data = JSON.parse(body);
                const apiKey = req.headers['x-api-key'] || req.headers['authorization']?.replace('Bearer ', '') || process.env.DEEPSEEK_API_KEY;
                
                if (!apiKey) {
                    res.writeHead(401, { 'Content-Type': 'application/json' });
                    res.end(JSON.stringify({ error: 'API key required' }));
                    return;
                }

                if (data.tools && Array.isArray(data.tools)) {
                    data.tools = data.tools.map(tool => {
                        if (tool.type === 'function' && tool.name && !tool.function) {
                            return {
                                type: 'function',
                                function: {
                                    name: tool.name,
                                    description: tool.description || '',
                                    parameters: tool.parameters || { type: 'object', properties: {} }
                                }
                            };
                        }
                        return tool;
                    });
                }

                if (data.model && THINKING_MODELS.includes(data.model)) {
                    if (!data.extra_body) {
                        data.extra_body = {};
                    }
                    if (!data.extra_body.thinking) {
                        data.extra_body.thinking = { type: "enabled" };
                    }
                    if (!data.reasoning_effort) {
                        data.reasoning_effort = "high";
                    }
                }

                const modifiedBody = JSON.stringify(data);
                const url = new URL(`${DEEPSEEK_BASE_URL}/v1/chat/completions`);

                const options = {
                    hostname: url.hostname,
                    port: 443,
                    path: url.pathname,
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json',
                        'Authorization': `Bearer ${apiKey}`
                    }
                };

                const proxyReq = https.request(options, (proxyRes) => {
                    res.writeHead(proxyRes.statusCode, proxyRes.headers);
                    proxyRes.pipe(res);
                });

                proxyReq.on('error', (err) => {
                    res.writeHead(500, { 'Content-Type': 'application/json' });
                    res.end(JSON.stringify({ error: err.message }));
                });

                proxyReq.write(modifiedBody);
                proxyReq.end();
            } catch (err) {
                res.writeHead(400, { 'Content-Type': 'application/json' });
                res.end(JSON.stringify({ error: err.message }));
            }
        });
        return;
    }

    res.writeHead(404, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ error: 'Not found', path: req.url }));
});

server.listen(PORT, '0.0.0.0', () => {
    console.log(`DeepSeek Gateway running on http://0.0.0.0:${PORT}`);
    console.log(`Proxying to: ${DEEPSEEK_BASE_URL}`);
    console.log(`Thinking mode enabled for: ${THINKING_MODELS.join(', ')}`);
    console.log('Endpoints:');
    console.log(`  POST http://localhost:${PORT}/v1/chat/completions`);
    console.log(`  GET  http://localhost:${PORT}/health`);
});

File: deepseek_gateway.py
#!/usr/bin/env python3
import os
import sys
import json
from http.server import HTTPServer, BaseHTTPRequestHandler
import urllib.request
import urllib.error

PORT = 5000
DEEPSEEK_BASE_URL = 'https://api.deepseek.com'
THINKING_MODELS = ['deepseek-v4-pro', 'deepseek-v4-flash']

class GatewayHandler(BaseHTTPRequestHandler):
    def log_message(self, format, *args):
        print(f"{self.log_date_time_string()} {args[0]}")

    def send_json_response(self, status_code, data):
        self.send_response(status_code)
        self.send_header('Content-Type', 'application/json')
        self.end_headers()
        self.wfile.write(json.dumps(data).encode())

    def do_GET(self):
        if self.path == '/health' or self.path == '/v1/models':
            self.send_json_response(200, {'status': 'ok'})
        else:
            self.send_json_response(404, {'error': 'Not found', 'path': self.path})

    def do_POST(self):
        if self.path == '/' or self.path.startswith('/v1/chat/completions'):
            content_length = int(self.headers.get('Content-Length', 0))
            body = self.rfile.read(content_length).decode('utf-8')
            
            try:
                data = json.loads(body)
            except json.JSONDecodeError as e:
                self.send_json_response(400, {'error': str(e)})
                return

            api_key = (
                self.headers.get('x-api-key') or 
                self.headers.get('Authorization', '').replace('Bearer ', '') or 
                os.environ.get('DEEPSEEK_API_KEY')
            )

            if not api_key:
                self.send_json_response(401, {'error': 'API key required'})
                return

            if data.get('tools') and isinstance(data['tools'], list):
                data['tools'] = [
                    {
                        'type': 'function',
                        'function': {
                            'name': tool['name'],
                            'description': tool.get('description', ''),
                            'parameters': tool.get('parameters', {'type': 'object', 'properties': {}})
                        }
                    } if tool.get('type') == 'function' and tool.get('name') and not tool.get('function') else tool
                    for tool in data['tools']
                ]

            if data.get('model') in THINKING_MODELS:
                if 'extra_body' not in data:
                    data['extra_body'] = {}
                if 'thinking' not in data['extra_body']:
                    data['extra_body']['thinking'] = {'type': 'enabled'}
                if 'reasoning_effort' not in data:
                    data['reasoning_effort'] = 'high'

            modified_body = json.dumps(data).encode('utf-8')

            req = urllib.request.Request(
                f'{DEEPSEEK_BASE_URL}/v1/chat/completions',
                data=modified_body,
                headers={
                    'Content-Type': 'application/json',
                    'Authorization': f'Bearer {api_key}'
                },
                method='POST'
            )

            try:
                with urllib.request.urlopen(req, timeout=120) as response:
                    self.send_response(response.status)
                    for key, value in response.getheaders():
                        self.send_header(key, value)
                    self.end_headers()
                    self.wfile.write(response.read())
            except urllib.error.HTTPError as e:
                self.send_response(e.code)
                self.wfile.write(e.read())
            except Exception as e:
                self.send_json_response(500, {'error': str(e)})
        else:
            self.send_json_response(404, {'error': 'Not found', 'path': self.path})

if __name__ == '__main__':
    print(f'DeepSeek Gateway running on http://0.0.0.0:{PORT}')
    print(f'Proxying to: {DEEPSEEK_BASE_URL}')
    print(f'Thinking mode enabled for: {", ".join(THINKING_MODELS)}')
    print('Endpoints:')
    print(f'  POST http://localhost:{PORT}/v1/chat/completions')
    print(f'  GET  http://localhost:{PORT}/health')
    
    server = HTTPServer(('0.0.0.0', PORT), GatewayHandler)
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print('\nShutting down...')
        server.shutdown()